src/borrow_tracker/tree_borrows/tree/tests.rs

```diff
@@ -0,0 +1,645 @@
+                // always ignore their value.
+                TestEvent::RetX(_) => write!(f, "ret x"),
+                TestEvent::RetY(_) => write!(f, "ret y"),
+                TestEvent::RetagY(newp) => write!(f, "retag y ({newp})"),
+            }
+        }
+    }
+
+    #[derive(Clone, PartialEq, Eq, Hash)]
+    /// The state of a pointer on a location, including the protector.
+    /// It is represented here with the protector bound to the `LocationState` rather
+    /// than the `Map<Location, LocationState>` as is normally the case,
+    /// but since all our exhaustive tests look at a  single location
+    /// there's no risk of `prot` for different locations of the same tag getting
+    /// out of sync.
+    struct LocStateProt {
+        state: LocationState,
+        prot: bool,
+    }
+
+    impl LocStateProt {
+        fn is_initial(&self) -> bool {
+            self.state.is_initial()
+        }
+
+        fn perform_access(&self, kind: AccessKind, rel: AccessRelatedness) -> Result<Self, ()> {
+            let mut state = self.state;
+            state.perform_access(kind, rel, self.prot).map_err(|_| ())?;
+            Ok(Self { state, prot: self.prot })
+        }
+
+        fn end_protector(&self) -> Self {
+            Self { prot: false, state: self.state }
+        }
+    }
+
+    impl Exhaustive for LocStateProt {
+        fn exhaustive() -> Box<dyn Iterator<Item = Self>> {
+            Box::new(
+                <(LocationState, bool)>::exhaustive().map(|(state, prot)| Self { state, prot }),
+            )
+        }
+    }
+
+    impl fmt::Display for LocStateProt {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "{}", self.state)?;
+            if self.prot {
+                write!(f, ", protect")?;
+            }
+            Ok(())
+        }
+    }
+
+    #[derive(Clone, PartialEq, Eq, Hash)]
+    /// The states of two pointers to the same location.
+    ///
+    /// Note that the two states interact: using one pointer can have
+    /// an impact on the other.
+    /// This makes `LocStateProtPair` more meaningful than a simple
+    /// `(LocStateProt, LocStateProt)` where the two states are not guaranteed
+    /// to be updated at the same time.
+    /// Some `LocStateProtPair` may be unreachable through normal means
+    /// such as `x: Active, y: Active` in the case of mutually foreign pointers.
+    struct LocStateProtPair {
+        xy_rel: RelPosXY,
+        x: LocStateProt,
+        y: LocStateProt,
+    }
+
+    impl LocStateProtPair {
+        fn perform_test_access(self, acc: &TestAccess) -> Result<Self, ()> {
+            let (xrel, yrel) = acc.ptr.rel_pair(self.xy_rel);
+            let x = self.x.perform_access(acc.kind, xrel)?;
+            let y = self.y.perform_access(acc.kind, yrel)?;
+            Ok(Self { x, y, ..self })
+        }
+
+        fn read_if_initialized(self, ptr: PtrSelector) -> Result<Self, ()> {
```

Comment:
I think this should panic if called on `Other`, it makes little sense there. Also it needs a doc comment.
```
/// Perform a read on the given pointer if its state is `initialized`.
/// Must be called just after reborrowing a pointer.
```